server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -160,4 +384,478 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+                        }
+                        for (int i = 0; i < response.results.length; i++) {
+                            var s = request.shards.get(i);
+                            int shardIdx = s.shardIndex;
+                            final SearchShardTarget target = new SearchShardTarget(routing.nodeId(), s.shardId, routing.clusterAlias());
+                            switch (response.results[i]) {
+                                case Exception e -> onShardFailure(shardIdx, target, shardIterators[shardIdx], e);
+                                case SearchPhaseResult q -> {
+                                    q.setShardIndex(shardIdx);
+                                    q.setSearchShardTarget(target);
+                                    onShardResult(q);
+                                }
+                                case null, default -> {
+                                    assert false : "impossible [" + response.results[i] + "]";
+                                }
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void handleException(TransportException e) {
+                        Exception cause = (Exception) ExceptionsHelper.unwrapCause(e);
+                        if (e instanceof SendRequestTransportException || cause instanceof TaskCancelledException) {
+                            // two possible special cases here where we do not want to fail the phase:
+                            // failure to send out the request -> handle things the same way a shard would fail with unbatched execution
+                            // as this could be a transient failure and partial results we may have are still valid
+                            // cancellation of the whole batched request on the remote -> maybe we timed out or so, partial results may
+                            // still be valid
+                            for (ShardToQuery shard : request.shards) {
+                                final int shardIndex = shard.shardIndex;
+                                onShardFailure(
+                                    shardIndex,
+                                    new SearchShardTarget(routing.nodeId(), shard.shardId, routing.clusterAlias()),
+                                    shardIterators[shardIndex],
+                                    e
+                                );
+                            }
+                        } else {
+                            // Remote failure that wasn't due to networking or cancellation means that the data node was unable to reduce
+                            // its local results. Failure to reduce always fails the phase without exception so we fail the phase here.
+                            if (results instanceof QueryPhaseResultConsumer queryPhaseResultConsumer) {
+                                queryPhaseResultConsumer.failure.compareAndSet(null, cause);
```

Comment:
do we need to do something if a failure was already set? What is the effect of this? Results that may be held are not returned, because an exception is set? But what if it was not an exception while reducing? Shouldn't this be handled differently on the data node side e.g. if there's an exception on merging in the data nodes can that be thrown as a specific exception? It seems like we are guessing a bit and making assumptions here otherwise?

Reply:
Yea we could make this cleaner, but as with other spots, it's more code and blows up the size of this thing. Essentially now (we had this on the coord side in another comment above) every non-cancellation failure that is thrown here counts as a hard failure of the whole phase. Adding a specific exception doesn't save much complication because it's either a cancellation or a hard stop here, but we still have to check for the send-transport exceptions on the coordinator anyway so that's why I left it as is for now.

Here too, I would suggest some follow-ups to move to chaining futures to make all of this a little easier to follow and reason about the safety of the code. The logic here is just to stay close to the original implementation for the time being really.